source/brailleInput.py

```diff
@@ -35,18 +41,239 @@ class BrailleInputHandler(object):
+		@type endWord: bool
+		@return: C{True} if translation produced text, C{False} if not.
+		@rtype: bool
 		"""
+		assert not self._table.contracted or endWord, "Must only translate contracted at end of word"
+		if self._table.contracted:
+			# self.bufferText has been used by _reportContractedCell, so clear it.
+			self.bufferText = u""
+		oldTextLen = len(self.bufferText)
+		pos = self.untranslatedStart + self.untranslatedCursorPos
 		# liblouis requires us to set the highest bit for proper use of dotsIO.
-		char = unichr(dots | 0x8000)
+		data = u"".join([unichr(cell | 0x8000) for cell in self.bufferBraille[:pos]])
+		self.bufferText = louis.backTranslate(
+			[os.path.join(brailleTables.TABLES_DIR, self._table.fileName),
+			"braille-patterns.cti"],
+			data, mode=louis.dotsIO)[0]
+		newText = self.bufferText[oldTextLen:]
+		if newText:
+			# New text was generated by the cells just entered.
+			if self._table.contracted:
+				# For contracted braille, an entire word is sent at once.
+				# Don't speak characters as this is sent.
+				speech._suppressSpeakTypedCharacters(len(newText))
+			self.untranslatedStart = pos
+			self.untranslatedCursorPos = 0
+			self.sendChars(newText)
+
+		if endWord:
+			# We only need to buffer one word.
+			# Clear the previous word (anything before the cursor) from the buffer.
+			del self.bufferBraille[:pos]
+			self.bufferText = u""
+			self.cellsWithText.clear()
+			self.untranslatedStart = 0
+			self.untranslatedCursorPos = 0
+
+		if newText or endWord:
+			self._updateUntranslated()
+			return True
+
+		return False
+
+	
+	BRAILLE_ZZ = [0b110101] * 2 # The letters "zz" in braille (dot 1 3 5 6 twice)
+	def _translateForReportContractedCell(self, pos):
+		"""Translate text for current input as required by L{_reportContractedCell}.
+		@return: The previous translated text.
+		@rtype: unicode
+		"""
+		# Append braille "zz" so full word contractions don't take effect.
+		cells = self.bufferBraille[:pos + 1] + self.BRAILLE_ZZ
+		# liblouis requires us to set the highest bit for proper use of dotsIO.
+		data = u"".join([unichr(cell | 0x8000) for cell in cells])
+		oldText = self.bufferText
 		text = louis.backTranslate(
-			[os.path.join(braille.TABLES_DIR, config.conf["braille"]["inputTable"]),
+			[os.path.join(brailleTables.TABLES_DIR, self._table.fileName),
 			"braille-patterns.cti"],
-			char, mode=louis.dotsIO)
-		chars = text[0]
-		if len(chars) > 0:
-			self.sendChars(chars)
+			data, mode=louis.dotsIO)[0]
+		# Strip the "zz".
+		self.bufferText = text[:-2]
+		return oldText
+
+	def _reportContractedCell(self, pos):
+		"""Report a guess about the character(s) produced by a cell of contracted braille.
+		It's not possible to report the exact characters because later cells might change text produced by earlier cells.
+		However, it's helpful for the user to have a rough idea.
+		For example, when typing "alw" for always in English, the user will hear a, l, w.
```

Comment:
Yes. In English contracted braille, "alw" is the contraction for "always". However, we don't know it will produce "always" until after the word is complete. So, as the user types, they will hear "a l w". I've tried to clarify this comment a bit, but it's a bit tricky to explain.